bin/usgscsm_cam_test.cc

```diff
@@ -1,74 +1,211 @@
-// - Load a CSM model state (stored in a .json file, just like
-//   an ISD model).
-// - Ability to export a CSM model in ISD format to a CSM model state file. 
+// - Test projecting rays from the camera to ground, then back,
+//   and compare with the original pixel values.  
 
-#include <usgscsm/UsgsAstroPlugin.h>
-#include <csm/RasterGM.h>
+#include <UsgsAstroPlugin.h>
+#include <RasterGM.h>
 #include <UsgsAstroLsSensorModel.h>
 
 #include <iostream>
-int main(int argc, char **argv) {
+#include <fstream>
+#include <getopt.h> // for parsing command-line options
+
+struct Options {
+  std::string model;              // the .json file in isd or model state format
+  std::string output_model_state; // the output model state in .json format
+  int sample_rate;
+  double subpixel_offset, height_above_datum;
+  Options(): sample_rate(0), subpixel_offset(0.0), height_above_datum(0.0) {}
+};
+  
+// Parse the input options with getopt.
+int parse_options(int argc, char **argv, Options & opt) {
+
+  // Collect the parsed options in a map
+  std::map<std::string, std::string> parsed_options;
+  
+  int digit_optind = 0;
+  while (1) {
+    int this_option_optind = optind ? optind : 1;
+    std::string short_options = ""; // no short options
+    int option_index = 0;
+    static struct option long_options[] = {
+      {"model",              required_argument, 0,  0},
+      {"output-model-state", required_argument, 0,  0},
+      {"sample-rate",        required_argument, 0,  0},
+      {"subpixel-offset",    required_argument, 0,  0},
+      {"height-above-datum", required_argument, 0,  0},
+      {"help",               no_argument,       0,  0}
+    };
+
+    int c = getopt_long(argc, argv, short_options.c_str(), long_options, &option_index);
+    if (c == -1) // done parsing all options
+      break;
+
+    std::string opt_name = long_options[option_index].name;
+    if (c == 0) {
+      if (optarg) {
+        // The option has a value
+        parsed_options[opt_name] = optarg;
+      } else {
+        // The option has no value
+        parsed_options[opt_name] = "";
+      }
+    } else {
+      // Something went wrong in parsing. The parser will print a message. Just add to it.
+      std::cout << "Cannot continue.\n";
+      return 1;
+    }
+  }
+
+  // Print unexpected input arguments
+  if (optind < argc) {
+    printf("Unexpected argument: ");
+    while (optind < argc)
+      printf("%s ", argv[optind++]);
+    printf("\n");
+
+    return 1;
+  }
+
+  // See if the user asked for help
+  bool print_help_and_exit = (parsed_options.find("help") != parsed_options.end());
+  
+  // It is safe to access non-existent values from a map, the result
+  // will be an empty string
+  opt.model = parsed_options["model"];
+  if (opt.model == "") 
+    print_help_and_exit = true;
 
-  if (argc != 2) {
-    std::cerr << "Usage: " << argv[0] << " <model file>" << std::endl;
+  // Enforce that the sample rate is an integer
```

Comment:
I don't think this is required. For example, you could test every half or tenth of a pixel. The for loops will need to be updated to doubles though.

Reply:
This block is no longer in the latest version I put up for review, per:  https://github.com/USGS-Astrogeology/usgscsm/pull/379

Reply:
I'm confused what the test step size has to do with #379. Can you clarify for me?

Reply:
I looked at your red text, but you were talking about the green text.

Reply:
It is hard to explain to the user what a sample rate of 10.23 is. Everybody understands what a sample rate of 10 is. If a subpixel offset is desired, that one is done with the --subpixel-offset option. So, if your sample rate is 10, and subpixel offset is 0.23, it will test pixels 10.23, 20.23, etc. What you are proposing amounts to testing pixels 10.23, 20.46, etc. I guess the end results are about the same, but with what I propose things may be easier to explain.

Reply:
Okay that sounds reasonable and the subpixel-offset makes more context this way.